File: Java/src/main/java/QuestionBank/Solution501_600/Solution589.java
package QuestionBank.Solution501_600;

import tool.Node;

import java.util.ArrayList;
import java.util.List;

/**
 * @author windows
 */
public class Solution589 {
    ArrayList<Integer> list = new ArrayList<>();
    public List<Integer> preorder(Node root) {
        if (root == null) {
            return list;
        }
        list.add(root.val);
        List<Node> children = root.children;
        if (children != null) {
            for (int i = 0; i < children.size(); i++) {
                preorder(children.get(i));
            }
        }
        return list;
    }


    public static void main(String[] args) {
        Solution589 solution589 = new Solution589();
    }
}

File: C++/src/Solution501_600/Solution589.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution589.h"
#include <functional>
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <string>
using namespace std;
#include <vector>
using namespace std;


// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

Node() {}

Node(int _val) {
  val = _val;
}

Node(int _val, vector<Node*> _children) {
  val = _val;
  children = _children;
}
};

void dfs(Node* root, vector<int>& list){
if (root == nullptr){
  return;
}
list.push_back(root->val);
for (const auto &item : root->children){
  dfs(item, list);
}
}

vector<int> preorder(Node* root) {
  vector<int> list;
  dfs(root, list);
  return list;
}
